### PR TITLE
[OpenCL] Add Unordered Queue

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

This is just a dummy PR for Rok, Sean, and myself to talk about the unordered event queue for the OpenCL code.

This was actually pretty easy:
1. In the [device queue](https://github.com/bstatcomp/math/compare/develop...bstatcomp:opencl-unordered-queue?expand=1#diff-23051186f20ae937b29d69ce3b413239R96) set the execution mode to `CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE`
2. Add a vector of events to each matrix_gpu, with a method for adding events to a matrices event vector.
3. For each kernel, write to device, and transfer within the device, we pass an event list to the kernel or method so that it knows what it has to wait on before it can go do stuff. These methods also pass back an event that we attach to the output matrix.
4. Keep each check and transfer back to the host as a blocking operation. This makes sure that everything is done or 'checkpointed' before the host sees anything.

The only thing I don't like about this is that the the signature for calling the kernels is getting a little bloated. 

```cpp
    cl::Event add_event = opencl_kernels::add(matrix_events,
      cl::NDRange(A.rows(), A.cols()), C.buffer(), A.buffer(),
      B.buffer(), A.rows(), A.cols());
```

Kinda gross looking

I think the easiest way to clean that up would be to make the `global_range_kernel` and `local_range_kernel` into a class. We could also shrink those down. I would imagine calling a kernel could look something like the below

```cpp
// setup the kernel
kernel_cl add = opencl_kernels::add(matrix_events, cl::NDRange(A.rows(), A.cols()));
// or maybe
kernel_cl add = opencl_kernels::add();
add.events(matrix_events);
add.global_range(cl::NDRange(A.rows(), A.cols());
// Then run
cl::Event add_event = add(C.buffer(), A.buffer(), B.buffer(), A.rows(), A.cols());
// Add event
C.events(add_event);
```

## Tests

I didn't want to rewrite everything until we had a final design so for now the only test that works with the OpenCL code is the of addition. Which is a low bar but does show that stuff works.

```bash
./runTests.py ./test/unit/math/gpu/add_test
```

## Side Effects

Yeah, I think this whole thing?

### CODE REVIEW

**Comment by SteveBronder** [comment]
@seantalts I can't add you as a reviewer here for some reason

**Comment by SteveBronder** [comment]
@rok-cesnovar can you make me an admin on this fork?

**Comment by rok-cesnovar** [comment]
I will have a look.

**Comment by rok-cesnovar** [comment]
You should have it now.

**Comment by seantalts** [comment]
Here is a good way to look at just this diff: https://github.com/bstatcomp/math/compare/gpu-cholesky-rev...bstatcomp:opencl-unordered-queue

Thanks for writing up all the considerations about how to use the API and putting up a couple prototypes. I think the first one is not too bad if you consider that we're going to always be wrapping opencl:: functions from within other Math functions, but it does seem like we should be able to abstract that away. In fact, I think we want to do that automatically - I think we can have `global_range_kernel` et al automatically add events that they find on their inputs. I'm only like 85% sure of that, but I think it's kind of a fun logic puzzle to see how to do it. You have to unpack the variadic `...Args` template arguments concat the events from each of them (and/or, as your comment suggests, write `event_concat` to take variadic args). How does that sound? Then we don't add the events arg to kernels, if that ends up working.

**Comment by SteveBronder** [comment]
Hmm, I think you are correct but I'm not sure if I'm getting what your throwing down. Let me try to break down how I'm interpreting the above

> I think we can have global_range_kernel et al automatically add events that they find on their inputs. I'm only like 85% sure of that, but I think it's kind of a fun logic puzzle to see how to do it. 

```c++
cl::Event add_event = add(C.buffer(), A.buffer(), B.buffer(), A.rows(), A.cols());
```

We would have to have the buffer return a struct of both the event vector and the actual buffer right?

If we go that route I'd prefer to abstract this even more and pass in the actual `matrix_gpu` objects

```c++
cl::Event add_event = add(C, A, B, A.rows(), A.cols());
```

We would then need prototypes in `matrix_cl` of the inplace methods and then declare those routines in headers later, but then I think there is some clever way we could parse out the events from the variadic template of `event_concat`

> You have to unpack the variadic ...Args template arguments concat the events from each of them (and/or, as your comment suggests, write event_concat to take variadic args). How does that sound? Then we don't add the events arg to kernels, if that ends up working.

If you agree with the above then I think we are both at that 85% probably possible level here.

**Comment by seantalts** [comment]
> If we go that route I'd prefer to abstract this even more and pass in the actual matrix_gpu objects

Yes I think this makes sense - I forgot we were always passing buffers. I think it makes sense even if we didn't have the events thing, no? Nice to have the global_range_kernel be calling the .buffer() on everything instead of all users of all kernels.

> prototypes in matrix_cl of the inplace methods 

prototypes of inplace methods? not sure what these words mean here haha. But everything else sounds good.

**Comment by SteveBronder** [comment]
> prototypes of inplace methods?

My b. Right now `matrix_cl` depends on `kernel_cl`. So to do the above we need to remove that dependency. For instance, with ` triangular_transpose()` I think we have to declare ` triangular_transpose()` in the matrix_cl class, but then define it in another header later. 

This feels more clear though and I think we are aligned. Let me make an example of that then we can chat about it

**Comment by SteveBronder** [comment]
I got this to work for everything! Staring at `nvtop` and running the rev test I saw throughput at about 30-50% ish higher than what I normally see

I need to clean this code up and clean I'll make a PR and post some tests

**Comment by rok-cesnovar** [comment]
Wow, that is awesome!

**Comment by seantalts** [comment]
That's awesome! Was the thing you were talking about these declarations here? https://github.com/bstatcomp/math/pull/9/files#diff-2336890ba7053d4cb33bf81074bb1a61R40 In C++ I think these are called "forward declarations" of the methods, though I don't think I've seen someone use them like this within its own class, I guess we have to because we're defining this class's methods in a bunch of different files. I wonder if we should not do that anymore - our closest equivalent is probably the `var` class which has everything in one file. I don't think there's a hard and fast rule either way.

**Comment by SteveBronder** [comment]
Yeah sorry I knew there was an actual word for  forward declarators but the error I got said something about prototypes so I goofed.

Yeah it's kinda weird, hmm maybe I can do the reverse here and forward declare the matrix_cl and its event.

I dont like how much boilerplate this adds, but it adds it at a pretty low level in a way I think is safest. I'll make an actual PR and ping you for review when this is ready

**Comment by SteveBronder** [comment]
@seantalts do you know another way we could abstract the boiler plate away? To be able to do

```cpp
cl::Event add_event = add(C, A, B, A.rows(), A.cols());
```

We have to pull out both the events and the buffer from the matrices. This is fine for the events since opencl wants a vector of events anyway. But I couldn't figure out how to unpack the buffers to go into the kernel functor. If you know a way to do that then we can whipe out all this boilerplate which would be awesome 

The only thing I saw was at the link below under 
'Function argument lists' I think you could have a weird template that actually pulls out the buffer and event to feed into the kernel funtor call. If that's possible then it becomes a question of whether abstract templating or direct boilerplate is better. If I can both get the templating to work and explain it that would be cool

https://en.cppreference.com/w/cpp/language/parameter_pack

**Comment by seantalts** [comment]
It's a little complicated (might not be worth the cleverness); I'll try to code up a prototype ...

File: stan/math/opencl/kernel_cl.hpp [hunk 1]
```diff
@@ -123,10 +169,11 @@ struct global_range_kernel {
    * @param args The arguments to pass to the kernel.
    * @tparam Args Parameter pack of all kernel argument types.
    */
-  auto operator()(cl::NDRange global_thread_size, Args... args) const {
-    auto f = make_functor();
-    cl::EnqueueArgs eargs(opencl_context.queue(), global_thread_size);
-    f(eargs, args...).wait();
+  template <typename... Margs>
```
Comment (seantalts) [inline stan/math/opencl/kernel_cl.hpp:172]: Why did you add this?

**Comment by seantalts** [comment]
Actually I think it's easier than I thought because of this: https://stackoverflow.com/questions/47848910/apply-function-on-each-element-in-parameter-pack

But we need to sort out the Margs thing - we don't need both Args and Margs, and I am not sure why we'd need to move the templating to the method just to get the buffers and events(?)

**Comment by SteveBronder** [comment]
Oh, yeah I did this at 2 am last night and was mostly fighting to get a working prototype. Yeah the Marg can go

Above link sort of makes sense. Would we have a `convert` function that for matrix_cl types returns the buffer and for everything else returns the input argument?

File: stan/math/opencl/kernel_cl.hpp (hunk 1 above)
Comment (SteveBronder) [reply stan/math/opencl/kernel_cl.hpp:172]: ¯\\_(ツ)_/¯

Hacky hacking, I'll get rid of it when I get home

**Comment by seantalts** [comment]
I think that'd be the easiest (possibly only) way to do it, yeah. basically have 
```cpp
template <typename T>
const T& maybe_get_buffer(const T& t) { return t; }

template <>
const cl::Buffer& maybe_get_buffer(const matrix_cl& m) { return m.buffer(); }
```

**Comment by seantalts** [comment]
Also I think we should change a bunch of stuff to `const &`, oops

**Comment by SteveBronder** [comment]
Yeah def, there's a lot of junk on this branch so I'm going to make a new one that cleans up everything

**Comment by seantalts** [comment]
Oh the const stuff is an older mistake of probably mine, haha :P Thinking specifically of the Args to the kernels - probably can all be references or maybe even const references? Anyway, not that important just saw it while looking at the kernel mechanisms again for the first time in a while.
